pkg/geo/geogfn/geohash.go

```diff
@@ -0,0 +1,193 @@
+// the whole feature. Larger features means less precision.
+func getPrecisionByBBox(box *geopb.BoundingBox) int {
+	minx := box.MinX
+	miny := box.MinY
+	maxx := box.MaxX
+	maxy := box.MaxY
+	precision := 0
+
+	// this is a point, for points we use the full precision
+	if minx == maxx && miny == maxy {
+		return 20
+	}
+
+	// Starts from a world bounding box:
+	lonmin := -180.0
+	latmin := -90.0
+	lonmax := 180.0
+	latmax := 90.0
+
+	// Each iteration shrinks the world bounding box by half, making
+	// adjustments each iteration until it intersects with
+	// the feature bbox.
+	for {
+		lonwidth := lonmax - lonmin
+		latwidth := latmax - latmin
+		latmaxadjust, lonmaxadjust, latminadjust, lonminadjust := 0.0, 0.0, 0.0, 0.0
+
+		// looks whether the longitudes of the bbox are to the left or
+		// the right of the world bbox longitudes, shrinks it and makes adjustments
+		// for the next iteration
+		if minx > lonmin+lonwidth/2.0 {
+			lonminadjust = lonwidth / 2.0
+		} else if maxx < lonmax-lonwidth/2.0 {
+			lonmaxadjust = -1 * lonwidth / 2.0
+		}
+		if lonminadjust != 0.0 || lonmaxadjust != 0.0 {
+			lonmin += lonminadjust
+			lonmax += lonmaxadjust
+			precision++
+		} else {
+			// we stop until the adjustment is 0, meaning the longitude
+			// intersected with the feature bounding box
+			break
+		}
+
+		// looks whether the latitudes of the bbox are to the left or
+		// the right of the world bbox latitudes, shrinks it and makes adjustments
+		// for the next iteration
+		if miny > latmin+latwidth/2.0 {
+			latminadjust = latwidth / 2.0
+		} else if maxy < latmax-latwidth/2.0 {
+			latmaxadjust = -1 * latwidth / 2.0
+		}
+		if latminadjust != 0.0 || latmaxadjust != 0.0 {
+			latmin += latminadjust
+			latmax += latmaxadjust
+			precision++
+		} else {
+			// we stop until the adjustment is 0, meaning the latitude
+			// intersected with the feature bounding box
+			break
+		}
+	}
+	// the precision (number of characters) is divided by 5
+	// since each geohash character is encoded as base32 (5 bits).
+	return precision / 5
+}
+
+// geohashPoint encodes the given latitude and longitude
+// by iterating downwards, gaining precision each iteration
+// until reaching the provided precision.
+// It encodes the latitude and the longitude into a binary string
+// The latitude is encoded as the odd bits and the longitude as the
+// even ones.
+// Every iteration (stopping until reaching the provided precision)
+// a range is used (starting from {-90.0, 90.0} for latitudes and
+// {-180.0, 180.0} for longitudes) and each bit represents whether the
+// latitude or longitude is to the left or to the right of that range.
+// Reaching 5 bits means we find the encoding character for that
+// precision level.
+// More info on: https://en.wikipedia.org/wiki/Geohash#Algorithm_and_example
+func geohashPoint(longitude float64, latitude float64, precision int) string {
+	isEven := true
+	i := 0
+	bitIdx := 0
+	bit := 0
+
+	// Initial lookup ranges
```

Comment:
For all these comments, let's begin with capital letters and end with full stops, e.g. "// Setup initial lookup ranges", "// Special base32 alphabet for geohashing".